src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1121,10 +1165,86 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
             falseTestPermitsTrueOther = true;
             trueTestImpliesTrueOther = false;
             falseTestImpliesTrueOther = false;
+            precondition = Tests.True.Instance;
+            sideeffect = Tests.True.Instance;
 
             // if the tests are for unrelated things, there is no implication from one to the other
-            if (!test.Input.Equals(other.Input))
-                return;
+            BoundDagTemp s1Input = test.Input;
+            BoundDagTemp s2Input = other.Input;
+            if (!s1Input.Equals(s2Input))
+            {
+                var preconditions = ArrayBuilder<Tests>.GetInstance();
+                do
+                {
+                    switch (s1Input.Source, s2Input.Source)
+                    {
+                        // Even though the two tests appear unrelated (with different inputs),
+                        // it is possible that they are in fact related under certain conditions.
+                        // For instance, the inputs [0] and [^1] point to the same element when length is 1.
+                        case (BoundDagIndexerEvaluation s1, BoundDagIndexerEvaluation s2):
+                            // Take the top-level input and normalize indices to account for indexer accesses inside a slice.
+                            // For instance [0] in nested list pattern [ 0, ..[$$], 2 ] refers to [1] in the containing list.
+                            (s1Input, BoundDagTemp s1LengthTemp, int s1Index) = GetCanonicalInput(s1);
+                            (s2Input, BoundDagTemp s2LengthTemp, int s2Index) = GetCanonicalInput(s2);
+                            Debug.Assert(s1LengthTemp.Syntax is ListPatternSyntax);
+                            Debug.Assert(s2LengthTemp.Syntax is ListPatternSyntax);
+                            // Ignore input source as it will be matched in the subsequent iterations.
+                            if (s1Input.IsEquivalentTo(s2Input) &&
+                                // We don't want to pair two indices within the same pattern.
+                                s1LengthTemp.Syntax != s2LengthTemp.Syntax)
+                            {
+                                Debug.Assert(s1LengthTemp.IsEquivalentTo(s2LengthTemp));
+                                Debug.Assert(s1LengthTemp.Equals(s2LengthTemp) == s1Input.Equals(s2Input));
+                                if (s1Index == s2Index)
+                                {
+                                    continue;
+                                }
```

Comment:
Nit: if these aren't `if/else`s, please add newlines for clarity (and below in this function).